odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+                                grad_lipschitz=(
+                                    np.abs(scalar) * func.grad_lipschitz))
+
+        elif scalar == 0:
+            Functional.__init__(self, domain=func.domain, linear=True,
+                                grad_lipschitz=0)
+
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
+
+        OperatorLeftScalarMult.__init__(self, operator=func, scalar=scalar)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.scalar * self.operator.gradient
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the scaled functional."""
+        # The helper function only allows positive scaling parameters.
+        # Otherwise it gives an error.
+        return ConvexConjugateFuncScaling(
+            self.operator.conjugate_functional, self.scalar)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the scaled functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+
+        sigma = float(sigma)
+
+        if self.scalar < 0:
+            raise ValueError('Proximal operator of functional scaled with a '
+                             'negative value {} is not well-defined'
+                             ''.format(self.scalar))
+
+        elif self.scalar == 0:
+            return IdentityOperator(self.domain)
+
+        elif self.scalar > 0:
+            return self.operator.proximal(sigma * self.scalar)
+
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'
+                            ''.format(self.scalar))
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.scalar * self.operator.derivative(point)
```

Comment:
For the function `g(x) = a*f(x)`, the derivative in the point `x` along direction `p` should be given by `<p, grad_g(x)> = <p, a*grad_f(x)>`, no? The `self.operator` is accessing the functional `f`, but I will rewrite that according to a previous comment.